lib/tree.rb

```diff
@@ -16,40 +16,168 @@ def initialize
   def find(key)
-    raise NotImplementedError
+    # raise NotImplementedError
+    current = @root
+    return nil if current.nil?
+    
+    if current.key == key
+      return current.value
+    else
+      find_helper(current, key)
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def find_helper(current, key)
+    return current.value if key == current.key
+
+    if key <= current.key
+      find_helper(current.left, key)
+    else
+      find_helper(current.right, key)
+    end
+  end
+  
+  # Me, Yieni and Haben got together to do the traverse methods.
+  # We found this link but worked on it for a few hours to understand
+  # http://rubyalgorithms.com/binary_search_tree.html
+  # whet &block and yield are doing before using them
+  
+  # Time Complexity: O(Log n)
+  # Space Complexity: O(n)
   def inorder
-    raise NotImplementedError
+    # raise NotImplementedError
+    root = @root
+    traverse_array = []
+    return traverse_array if root.nil?
+
+    inorder_helper(root) do |node|
+      traverse_array << { key: node.key, value: node.value }
+    end
+
+    traverse_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # any method in ruby can take a block
+  # &block is usuful for sending in the block when we want to pass it on
+  # but with yield alone we did not need to pass in the &block into the main
+  # inorder_helper method
+  # if we want to use block.call(node) we need to pass in &block in the main
+  # helper method
+
+  # yield returns the variable - the information back up :)
+  
+  def inorder_helper(node, &block)
+    return if node.nil?
+
+    inorder_helper(node.left, &block)
+    yield node
+    inorder_helper(node.right, &block)
+  end
+
+  # Time Complexity: O(Log n)
+  # Space Complexity: O(n)
   def preorder
-    raise NotImplementedError
+    # raise NotImplementedError
+    root = @root
+    traverse_array = []
+    return traverse_array if root.nil?
+
+    preorder_helper(root) do |node|
+      traverse_array << { key: node.key, value: node.value }
+    end
+
+    traverse_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def preorder_helper(node, &block)
+    return if node.nil?
+    
+    yield node
+    preorder_helper(node.left, &block)
+    preorder_helper(node.right, &block)
+  end
+
+  # Time Complexity: O(Log n)
+  # Space Complexity: O(n)
   def postorder
-    raise NotImplementedError
+    # raise NotImplementedError
+    root = @root
+    traverse_array = []
+    return traverse_array if root.nil?
+
+    postorder_helper(root) do |node|
+      traverse_array << { key: node.key, value: node.value }
+    end
+
+    traverse_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def postorder_helper(node, &block)
+    return if node.nil?
+    
+    postorder_helper(node.left, &block)
+    postorder_helper(node.right, &block)
+    yield node
+  end
+
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def height
```

Comment:
⚠ This doesn't quite work as it goes down the left and right sides of the tree but doesn't investigate all paths to the leaves. 

A better solution:

```ruby
def height_helper(current_node)
    return 0 if current_node.nil?
    return 1 +[height_helper(current_node.left), height_helper(current_node.right)].max
end
```